clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -0,0 +1,183 @@
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.GSSManager;
+import org.ietf.jgss.GSSName;
+import org.ietf.jgss.Oid;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+import org.apache.kafka.common.network.Authenticator;
+import org.apache.kafka.common.network.NetworkSend;
+import org.apache.kafka.common.network.NetworkReceive;
+import org.apache.kafka.common.network.TransportLayer;
+import org.apache.kafka.common.security.auth.PrincipalBuilder;
+
+public class SaslServerAuthenticator implements Authenticator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SaslServerAuthenticator.class);
+
+    private final SaslServer saslServer;
+    private final Subject subject;
+    private final String node;
+
+    // assigned in `configure`
+    private TransportLayer transportLayer;
+
+    // buffers used in `authenticate`
+    private NetworkReceive netInBuffer;
+    private NetworkSend netOutBuffer;
+
+    public SaslServerAuthenticator(String node, final Subject subject) throws IOException {
+        this.node = node;
+        this.subject = subject;
+        saslServer = createSaslServer();
+    }
+
+    public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder, Map<String, ?> configs) {
+        this.transportLayer = transportLayer;
+    }
+
+    private SaslServer createSaslServer() throws IOException {
+        if (subject != null) {
+            // server is using a JAAS-authenticated subject: determine service principal name and hostname from kafka server's subject.
+            if (subject.getPrincipals().size() > 0) {
+                try {
+                    final SaslServerCallbackHandler saslServerCallbackHandler = new SaslServerCallbackHandler(Configuration.getConfiguration());
+                    final Principal servicePrincipal = subject.getPrincipals().iterator().next();
+
+                    final String servicePrincipalNameAndHostname = servicePrincipal.getName();
+                    int indexOf = servicePrincipalNameAndHostname.indexOf("/");
+                    final String servicePrincipalName = servicePrincipalNameAndHostname.substring(0, indexOf);
+                    final String serviceHostnameAndKerbDomain = servicePrincipalNameAndHostname.substring(indexOf + 1, servicePrincipalNameAndHostname.length());
+                    indexOf = serviceHostnameAndKerbDomain.indexOf("@");
+                    final String serviceHostname = serviceHostnameAndKerbDomain.substring(0, indexOf);
+                    final String mech = "GSSAPI";
+
+                    LOG.debug("serviceHostname is '" + serviceHostname + "'");
+                    LOG.debug("servicePrincipalName is '" + servicePrincipalName + "'");
+                    LOG.debug("SASL mechanism is '" + mech + "'");
+                    boolean usingNativeJgss = Boolean.getBoolean("sun.security.jgss.native");
```

Comment:
Could we make sun.security.jgss.native a property in the broker/client config file? In general, it seems that other than the jaas config file, it's better to specify other properties from config file instead of system properties.